src/usdb_syncer/gui/mw.py

```diff
@@ -263,6 +269,77 @@ def _generate_song_pdf(self) -> None:
         if path:
             generate_song_pdf(self.table.all_local_songs(), path)
 
+    def _import_usdb_ids_from_files(self) -> None:
+        file_list = QFileDialog.getOpenFileNames(
+            self,
+            caption="Select one or more files to import USDB IDs from",
+            dir=os.getcwd(),
+            filter="JSON, USDB IDs, Weblinks (*.json *.usdb_ids *.url *.webloc *.desktop)",
+        )[0]
+        if not file_list:
+            _logger.info("no files selected to import USDB IDs from")
+            return
+        file_parsers = [UsdbIdFile.parse(path) for path in file_list]
```

Comment:
I still don't understand the purpose of this class. I would simplify things as follows:
```py
song_ids = []
for path in file_list:
    try:
        song_ids += parse_usdb_id_file(path)
    except UsdbIdFileError as error:
        ... # log error
        return
```

`parse_usdb_id_file()` would be the new `UsdbIdFile.parse()`, but simply returning ids (or raising).
Likewise you would define a function `write_usdb_id_file()` that only takes a list of ids. The current `UsdbIdFile.write()` takes an instance that might also represent an error, which is a bit weird.

Reply:
IMHO it felt correct to represent the file with an object rather than a function. To be honest my goals was to avoid exception handling outside in mw.py, as it would require more lines to differentiate between success and failure of individual files. Also we might want to extend the functionality at some point in future (e.g. exporting with artist and title or some flags). Further I thought that writing might also raise exceptions (e.g. file blocked for writing or invalid path) which are currently not handled.

Anyways I understand your point and will change the code accordingly. I just have to make up my mind how to deal with the exceltions in mw.py then. so that a single bad file does not block the import of several files.

Reply:
I appreciate that you're trying to keep code of mw.py. You can still do the exception handling in usdb_id_file.py if you add another function that wraps `parse_usdb_id_file()` and does the logging.
If you also pass it a song table reference, it can do the id validation as well.

If you think UsdbIdFile might be extended in the future, that's a fair point. I don't mind this class in general, just its current form: It's basically a union type that can represent a file _or_ a parse failure. That's confusing and cumbersome to work with in my opinion.
If the class simply represented a valid id file, I wouldn't mind it.